devito/yask/data.py

```diff
@@ -0,0 +1,311 @@
+import sys
+
+import numpy as np
+
+from devito.logger import yask as log
+from devito.tools import as_tuple
+
+from devito.yask.utils import rawpointer
+
+
+class Data(object):
+
+    """
+    A ``Data`` wraps a YASK grid.
+
+    A ``Data`` implements a subset of the ``numpy.ndarray`` API. The subset of
+    API implemented should suffice to transition between Devito backends w/o changes
+    to the user code. It was not possible to subclass ``numpy.ndarray``, as this
+    would have led to shadow data copies, since YASK employs a storage layout
+    different than what users expect (row-major).
+
+    The storage layout of a YASK grid looks as follows: ::
+
+        --------------------------------------------------------------
+        | extra_padding | halo |              | halo | extra_padding |
+        ------------------------    domain    ------------------------
+        |       padding        |              |       padding        |
+        --------------------------------------------------------------
+        |                         allocation                         |
+        --------------------------------------------------------------
+
+    :param grid: The YASK yk::grid that will be wrapped. Data storage will be
+                 allocated if not yet allocated.
+    :param shape: Shape of the domain region in grid points.
+    :param dimensions: A tuple of :class:`Dimension`s, representing the dimensions
+                       of the ``Data``.
+    :param halo: An integer indicating the extent of the halo region.
+    :param dtype: A ``numpy.dtype`` for the raw data.
+
+    .. note::
+
+        This type supports logic indexing over modulo buffered dimensions.
+    """
+
+    # Force __rOP__ methods (OP={add,mul,...) to get arrays, not scalars, for efficiency
+    __array_priority__ = 1000
+
+    def __init__(self, grid, shape, dimensions, halo, dtype):
```

Comment:
Note for reviewers: this is the only thing that really changed: the construction of the object, which mimics what numpy does with "main" arrays and views; just like numpy arrays, the attribute ``base`` stores a reference to the main array (ie, ``self.base != None`` ==> ``self`` is a view of ``base``)